core/http_component.cxx

```diff
@@ -182,20 +258,91 @@ class http_component_impl
       session = std::move(s);
     }
 
-    auto op = std::make_shared<pending_http_operation>(io_, request);
-    op->start(
-      [callback = std::move(callback)](auto resp, auto ec) mutable {
-        callback(std::move(resp), ec);
-      },
-      [session_manager, session, service = request.service]() mutable {
-        session_manager->check_in(service, session);
-      });
-    op->send_to(session);
+    op->start([callback = std::move(callback)](auto resp, auto ec) mutable {
+      callback(std::move(resp), ec);
+    });
+    op->set_stream_end_callback([session_manager, session, service = request.service]() mutable {
+      session_manager->check_in(service, session);
+    });
 
+    if (!session->is_connected()) {
+      session_manager->connect_then_send_pending_op(
+        session,
+        {},
+#ifdef COUCHBASE_CXX_CLIENT_COLUMNAR
+        op->dispatch_deadline_expiry(),
+#endif
+        op->deadline_expiry(),
+        [op = std::move(op)](std::error_code ec, std::shared_ptr<io::http_session> http_session) {
+          if (ec) {
+            return op->invoke_response_handler(ec, {});
+          }
+          op->send_to(http_session);
+        });
+    } else {
+      op->send_to(session);
+    }
     return op;
   }
 
 private:
+#ifdef COUCHBASE_CXX_CLIENT_COLUMNAR
+  auto defer_command(std::shared_ptr<pending_http_operation> pending_op,
+                     std::shared_ptr<io::http_session_manager> session_manager,
+                     const couchbase::core::cluster_credentials& credentials,
+                     free_form_http_request_callback&& callback)
+    -> tl::expected<std::shared_ptr<pending_operation>, std::error_code>
+  {
+    if (session_manager->last_bootstrap_error().has_value()) {
+      return tl::unexpected(session_manager->last_bootstrap_error().value().ec);
+    }
+    pending_op->start([callback = std::move(callback)](auto resp, auto ec) mutable {
+      callback(std::move(resp), ec);
+    });
+    // TODO:  client_context_id?
+    CB_LOG_DEBUG(R"(Adding pending HTTP operation to deferred queue: {})",
+                 pending_op->request().service);
+    session_manager->add_to_deferred_queue([op = std::move(pending_op),
```

Comment:
Similar here, `pending_op` shouldn't be moved here as it's being returned from this method.

Reply:
Ahh, thanks!  Will update 👍